zebra-state/src/service.rs

```diff
@@ -565,9 +573,16 @@ impl StateService {
         tracing::debug!(block = %prepared.block, "queueing block for contextual verification");
         let parent_hash = prepared.block.header.previous_block_hash;
 
-        if self.mem.any_chain_contains(&prepared.hash)
-            || self.read_service.db.hash(prepared.height).is_some()
+        if self
+            .sent_non_finalized_block_hashes
+            .contains(&prepared.hash)
         {
+            let (rsp_tx, rsp_rx) = oneshot::channel();
+            let _ = rsp_tx.send(Err("block already sent to be committed to the state".into()));
+            return rsp_rx;
+        }
+
+        if self.read_service.db.hash(prepared.height).is_some() {
```

Comment:
I opened PR #5323 with a faster way of implementing this.

(RocksDB has a "key is present" and a "get the data" API.)

Reply:
that seems better.